### ISSUE TITLE
Unable to upload via Table.shrine_field = File.open(...)

### ISSUE THREAD

**Issue by hms** [janko-m/shrine#122]
TL;DR version:
I've been trying to upload test data by: 
`table.shrine_field = File.open(path_to_file)`

I get a notification that the file has been cached and I can verify that the file has, in fact, been uploaded to the cache location.  But the file never makes it to the store location and I get a file closed error.

I've pulled out my shrine configuration out of my application and into a standalone minimal Rails test app with the following configuration.

Shrine initializer:
```
require 'shrine/storage/s3'

Shrine.plugin :activerecord
Shrine.plugin :direct_upload
Shrine.plugin :restore_cached_data
Shrine.plugin :cached_attachment_data
Shrine.plugin :determine_mime_type
Shrine.plugin :logging
Shrine.plugin :backgrounding

s3_options = {
  access_key_id: ENV['S3_ACCESS_KEY_ID'],
  secret_access_key: ENV['S3_SECRET_ACCESS_KEY'],
  region: ENV['S3_REGION'],
  bucket: ENV['S3_BUCKET']
}

Shrine.storages = {
  cache: Shrine::Storage::S3.new(prefix: 'stage', **s3_options),
  store: Shrine::Storage::S3.new(**s3_options)
}
```

The model:
```
class Test < ApplicationRecord
  include VideoUploader[:video]
end
```

The uploader:
`class VideoUploader < Shrine; end`

The database table:
Integer: id
Text: video

My best guess is it's something with the interaction and the latest aws-sdk, but I can't seem to lock things down as to where it's happening.  I get as far as this line:
```
def shrine_class
   self.class.shrine_class
end
```

But can't seem to figure out what's triggering the failure.

**Comment by janko-m** [comment]
Can you include the backtrace of the error, as well as your form and the controller code which is assigning the file?

**Comment by hms** [comment]
@janko-m That's the weird thing -- I'm not getting a backtrace.  It looks as if code somewhere is catching the error and just printing a message (and damned if I can find out where). 

Also, I'm not running this through the view/controller layer -- I'm doing this all via the command line.
```ruby
v = Video.create(...)
v.shrine = File.open(...)
```

I've been trying to single step the code to get a better idea of what's going on using both simple pry debugger and rubymines visual debugger without much luck. I've also tried to streamline my other development gems to see if one of them is swallowing an exception.   My next step is to create a minimal environment with just Rails, pg, aws-sdk, roda, and shrine (I think that's all I'll need), and see if I can reproduce the issue.

Thanks for all of the help for the past couple of weeks.

Hal

**Comment by hms** [comment]
I've created a small sample that I believe is the smallest chunk of code that demonstrates the problem.

This code is pretty much:
rails new .
Adding shrine to the Gemfile
creating a single table with the shrine_data field
creating an empty AR class
creating a minimal Shrine initializer
creating a minimal VideoUploader

The example that fails is:
```
v = Video.create
v.shrine = File.new('test file')
```

When I run this, I get:
Video.first.shrine = File.new('../sign_off/test_data/videos/noah/small.m4v')
  Video Load (0.3ms)  SELECT  "videos".* FROM "videos" ORDER BY "videos"."id" ASC LIMIT $1  [["LIMIT", 1]]
2016-12-24T00:54:27.833Z 9489: STORE[cache] VideoUploader[:shrine] Video[1] 1 file (5.98s)
 => #<File:video.m4v (closed)> 

Here is all of the code, minus a .env file with the S3 info (it should match format you used for your test environment).

[example.tar.gz](https://github.com/janko-m/shrine/files/671596/example.tar.gz)

Hal

**Comment by janko-m** [comment]
Thanks for creating a minimal app, though I'm still not entirely clear what is the behaviour that you expect, and how does the actual behaviour differ from the expected.

Shrine always closes the uploaded files, that's documented behaviour. The only way that you should get a "closed file" error is if you try to upload the same file twice.

I would be great if you post a self-contained script with `require "./config/environment"` on the top (which loads the Rails environment), which is run from the root of the Rails app, and you show what error does that script produce. That way I can run it as well, and easily debug it it turns out to be a bug.

**Comment by janko-m** [comment]
This is a basic example which definitely works (otherwise Shrine would be broken), so I will close this until it's proven to be a bug.
